sections/security/expirejwt.md

```diff
@@ -2,43 +2,55 @@
 
 ### One Paragraph Explainer
 
-By design, JWTs (JSON Web Tokens) are completely stateless, so once a valid token is signed by an issuer, the token may be verified as authentic by the application. The problem this leads to is the security concern where a leaked token could still be used and unable to be revoked, due to the signature remaining valid as long as the signature provided by the issues matches what the application is expecting.
-Due to this, when using JWT authentication, an application should manage a blacklist of expired or revoked tokens to retain user's security in the case a token needs to be revoked.
+By design, JWTs (JSON Web Tokens) are completely stateless, so once a valid token is signed by an issuer, the token may be verified as authentic by the application. The problem this leads to is the security concern where a leaked token could still be used and be unable to be revoked, due to the signature remaining valid as long as the signature provided by the issues matches what the application is expecting.
+Due to this, when using JWT authentication, an application should manage a denylist of expired or revoked tokens to retain user's security in the case a token needs to be revoked.
 
-### `express-jwt-blacklist` example
+### Redis example
 
-An example of running `express-jwt-blacklist` on a Node.js project using the `express-jwt`. Note that it is important to not use the default store settings (in-memory) cache of `express-jwt-blacklist`, but to use an external store such as Redis to revoke tokens across many Node.js processes.
+An example of using Nodejs and a Redis store containing your denylist. Keeping your denylist in an array or object is not recommended since if you restart, you'd lose it, and other instances wouldn't know if someone was explicitly logged out. So this assumes you have a running instance of Redis or some persistent key-value store and middleware handling login/JWT creation, and JWT verification since this is also not an extensive example of JWT implementation.
 
 ```javascript
-const jwt = require('express-jwt');
-const blacklist = require('express-jwt-blacklist');
-
-blacklist.configure({
-  tokenId: 'jti',
-  strict: true,
-  store: {
-    type: 'memcached',
-    host: '127.0.0.1',
-    port: 11211,
-    keyPrefix: 'mywebapp:',
-    options: {
-      timeout: 1000
+const redis = require("redis");
+const redisClient = redis.createClient();
+
+// Make sure you handle the creation and verification of the JWT before checking its logout status
+
+// Check for logout middleware
+app.use((req,res,next) => {
+   const { userId, tokenIssuedAt } = req;
+
+  // Using the token itself as the matching key is not reliable, so use the content of the JWT

```

Comment:
From my comment on the issue: https://github.com/goldbergyoni/nodebestpractices/issues/955#issuecomment-821274849
Referencing the comment on the blog post: https://dev.to/hessman_/comment/1c9h9
Valid base64 encoded strings can have padding of "=" at the end. So one could maliciously log in with an expired token by just adding padding, since it would otherwise be a valid token.

Reply:
Very interesting

Reply:
I wasn't sure if going into detail about that vulnerability would be extraneous. I'm happy to add more about it, since I suppose that might effect other things people would want to do with JWTs and would be good to know about regardless of whether they're implementing a denylist.

Yes that's it. Some suggest using the decoded signature as the matching key, but it's not text encoded as base64, so you'll probably need to convert it to some other format to store it. Though I just realized you could compare them as buffers using `Buffer.compare()`, so as long as you can reliably store the token as derived from binary/buffer format, then you could compare the whole tokens (though you'd need to remove the `.` separators for it to be valid base64). So perhaps the best bet is to just lop off the end segment `jwtString.split('.')[2]` and store and compare that using the base64 parsed to a buffer (which should take care of any padding.

Reply:
Very interesting